### PR TITLE
Enable cppcoreguidelines-special-member-functions

### PR DESCRIPTION
**Pull request by cyyever**
Fixes #ISSUE_NUMBER

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @xmfan @yf225

### CODE REVIEW

File: torch/csrc/profiler/collection.cpp [hunk 1]
```diff
@@ -402,6 +402,10 @@ struct StealOrDefault {
   explicit StealOrDefault(T& container)
       : container_{container}, it_{container.begin()} {}
 
+  StealOrDefault(const StealOrDefault&) = delete;
```
Comment (sraikund16) [inline torch/csrc/profiler/collection.cpp:405]: are these specific classes called out because they are marked as explicit?

**Review by sraikund16 — APPROVED** [review]
overall, lgtm

File: torch/csrc/profiler/collection.cpp (hunk 1 above)
Comment (cyyever) [reply torch/csrc/profiler/collection.cpp:405]: Because the member is reference and they are not used. The easiest way is to delete them.

**Pull request closed without merging**